### PR TITLE
fix: validate lower bound for dependency indices in upb file_def

### PR DESCRIPTION
**Pull request by KevinZhao**
## Summary

- Add lower-bound (`< 0`) validation for `public_dependency` and `weak_dependency` indices in `_upb_FileDef_Create` (`upb/reflection/file_def.c`)
- The existing checks only validated the upper bound (`>= dep_count`). Since both the index (`int32_t`) and `dep_count` (`int`) are signed, negative values like `-1` pass the comparison (`-1 >= 3` is false in signed arithmetic), get stored, and cause an out-of-bounds read when later used to index the `deps` array via `upb_FileDef_PublicDependency()` or `upb_FileDef_WeakDependency()`
- Add regression tests for negative `public_dependency` and `weak_dependency` values
- Add a fuzz target (`file_def_fuzz_test.cc`) that exercises the dependency accessor code paths after `DefPool.AddFile()`, covering index validation gaps that the existing `def_to_proto_fuzz_test` does not reach

## Test plan

- [ ] `bazel test //upb/reflection:reflection_test` — new tests verify negative indices are rejected with "out of range" error
- [ ] `bazel test //upb/reflection:def_builder_test` — existing tests still pass
- [ ] Manual verification: before this fix, `public_dependency: -1` on a file with dependencies causes SIGSEGV in `upb_FileDef_PublicDependency()`; after the fix, `_upb_FileDef_Create` rejects it with an error

### CODE REVIEW

**Comment by esrauchg** [comment]
Thanks for the candidate change!

Looks like the unit test fails for a different reason than the expected case, can you fix?

**Comment by KevinZhao** [comment]
Thanks @esrauchg! Fixed in the latest two commits:

**Commit 1** — Fixed the test: removed `dependency: "dep.proto"` from the test protos. The dependency name resolution (line 369-383) runs before the index validation, so the unloaded dep caused a different error than expected.

**Commit 2** — While reviewing I also found a pre-existing bug: `upb_FileDef_WeakDependency()` at line 145 asserts against `public_dep_count` instead of `weak_dep_count`. Fixed that too, and added boundary tests for `index=0` with `dep_count=0` (upper-bound OOB).

The "safe for tests" tag is needed for the fork CI to run the full test suite.

File: upb/reflection/file_def_fuzz_test.cc [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include "google/protobuf/descriptor.pb.h"
+#include "google/protobuf/descriptor.upb.h"
+#include <gtest/gtest.h>
+#include "testing/fuzzing/fuzztest.h"
+#include "upb/base/status.hpp"
+#include "upb/mem/arena.hpp"
+#include "upb/reflection/def.h"
+#include "upb/reflection/def.hpp"
+
+namespace upb_test {
```
Comment (esrauchg) [inline upb/reflection/file_def_fuzz_test.cc]: Do you mind sending a separate PR with this fuzz test? I'd like another maintainer to review this part, but I'm comfortable approving the other two files of this PR

Thanks!

**Review by esrauchg — CHANGES_REQUESTED** [review]
Thank you for the contribution!

**Comment by KevinZhao** [comment]
@esrauchg Done — split the fuzz test into a separate PR: #26720

This PR now contains only the two files you're comfortable approving:
- `upb/reflection/file_def.c` — lower-bound validation fix + WeakDependency assert fix
- `upb/reflection/reflection_test.cc` — unit tests for negative indices and zero-index boundary cases

Squashed into a single clean commit for easy review.

**Comment by KevinZhao** [comment]
Hi @esrauchg — the fuzz test has been split into a separate PR (#26720) as requested. This PR now only contains the two files you mentioned being comfortable with:

- `upb/reflection/file_def.c` — lower-bound validation fix + WeakDependency assert fix
- `upb/reflection/reflection_test.cc` — unit tests for boundary cases

CLA check is also passing now. Could you re-review when you get a chance? Thanks!

**Review by esrauchg — APPROVED** [review]

**Review by rgoldfinger6 — APPROVED** [review]
